stylix/mk-target.nix

```
@@ -0,0 +1,161 @@
+#           }
+#         )
+#
+#         (
+#           { fonts }:
+#           {
+#             programs.example.font.name = fonts.monospace.name;
+#           }
+#         )
+#       ];
+#     }
+#
+# Note that `mkTarget` is provided in the same place as module arguments.
+#
+# Ideally, in the future, this function also prevents unguarded module arguments
+# by limiting the arguments passed through `/stylix/autoload.nix`
+{
+  # The target name used to generate options in the stylix.targets.${name}
+  # namespace.
+  name,
+
+  # The descriptive target name passed to the lib.mkEnableOption function when
+  # generating the stylix.targets.${name}.enable option.
+  humanName,
+
+  # Whether the target should be automatically enabled by default according to
+  # the stylix.autoEnable option.
+  #
+  # This should be disabled if manual setup is required or if auto-enabling
+  # causes issues.
+  autoEnable ? true,
+
+  # Additional options to be added in the stylix.targets.${name} namespace along
+  # the stylix.targets.${name}.enable option.
+  #
+  # For example, an extension guard used in the configuration can be declared as
+  # follows:
+  #
+  #     { extension.enable = lib.mkEnableOption "the bloated dependency"; }
+  extraOptions ? { },
+
+  # Configuration functions that are automatically safeguarded when any of their
+  # arguments is disabled. The provided 'cfg' argument conveniently aliases to
+  # config.stylix.targets.${name}.
+  #
+  # For example, the following configuration is not merged if the stylix.colors
+  # option is disabled:
+  #
+  #     (
+  #       { colors }:
+  #       {
+  #         programs.example.theme.background = colors.base00;
+  #       }
+  #     )
+  #
+  # The 'cfg' alias can be accessed as follows:
+  #
+  #     (
+  #       { cfg }:
+  #       {
+  #         programs.example.extension.enable = cfg.extension.enable;
+  #       }
+  #     )
+  configElements ? [ ],
+
+  # This argument mirrors the configElements argument but intentionally lacks
+  # automatic safeguarding and should only be used for complex configurations
+  # where configElements is unsuitable.
+  generalConfig ? null,
+}:
+# Note: we only return a plain module so that calling code can use module args,
+# while using this as their module body
+let
+  module =
+    { config, lib, ... }:
+    let
+      cfg = config.stylix.targets.${name};
+
+      # Get the list of function de-structured argument names.
+      functionArgNames =
+        fn:
+        lib.pipe fn [
+          lib.functionArgs
+          builtins.attrNames
+        ];
+
+      getStylixAttrs =
+        fn:
+        lib.genAttrs (functionArgNames fn) (
+          arg:
+          if arg == "cfg" then
+            cfg
+          else if arg == "colors" then
+            config.lib.stylix.colors
+          else
+            config.stylix.${arg}
+              or (throw "stylix: Stylix argument requested by mkTarget does not exist: ${arg}")
```

Would this flow better with the name inline?

```suggestion
              or (throw "stylix: Stylix argument `${arg}` requested by mkTarget does not exist")
```

---

Is it worth printing the available names?

```suggestion
              or (throw "stylix: Stylix argument `${arg}` requested by mkTarget does not exist. Expected one of `cfg`, `colors`, ${
                lib.concatMapStringsSep ", " (name: "`${name}`") (builtins.attrNames config.stylix)
              }")
```

If you'd rather quote the names instead of backticks, you can use `builtins.toJSON` to render a quoted string.

---

> Would this flow better with the name inline?
>
> ```suggestion
>               or (throw "stylix: Stylix argument `${arg}` requested by mkTarget does not exist")
> ```

Although this flows better, it makes the message non-trivially parsable. When the message has only one parameter, adding it at the end with a `: ` field separator is the convention.

> Is it worth printing the available names?

Yes, informative error messages are always appreciated.

> ```suggestion
>               or (throw "stylix: Stylix argument `${arg}` requested by mkTarget does not exist. Expected one of `cfg`, `colors`, ${
>                 lib.concatMapStringsSep ", " (name: "`${name}`") (builtins.attrNames config.stylix)
>               }")
> ```

What about the "expected <EXPECTATION>, got <REALITY>" format:

> ```suggestion
>               or (throw "stylix: mkTarget expected one of `cfg`, `colors`, ${
>                 lib.concatMapStringsSep
>                 ", "
>                 (name: "`${name}`")
>                 (builtins.attrNames config.stylix)
>               }, got ${arg}")
> ```

> If you'd rather quote the names instead of backticks, you can use `builtins.toJSON` to render a quoted string.

I would also be fine without any quotation. But either way is fine for me.